### PR TITLE
feat(bundler): add config command for hyphen hosts

### LINKED ISSUE #9282: Bundler authentication should use `bundle config` instead of env variables
Renovate supports private bundler repositories: https://docs.renovatebot.com/modules/manager/bundler/#authenticating-private-registry

The implementation uses env variables to pass the credentials to the docker child container.
I have the use case that our internal repository has a dash in the name.

The resulting env var would be `BUNDLE_ARTIFACTORY-CORP__COM=user:password`.
This breaks bash as dashes are not allowed.
Source: https://bundler.io/v1.16/bundle_config.html#CREDENTIALS-FOR-GEM-SOURCES

Therefore the authentication should not happen with env vars but with `bundle config artifactory-corp.com user:password` before the bundler command is run

**Did you already have any implementation ideas?**

Yes, we can simply rework the already existing implementation to use the bundle config way instead of the env vars.
The effect is the same but we do now support hosts with dashes in the name.

We only need to mask the credentials from the log which should be no problem.

If this is fine with you I can implement it.

### PR DESCRIPTION
**Pull request by Chumper**
## Changes:

Add:
* `bundler config set --local <url> <creds>` for Bundler >= 2   
* `bundler config --local <url> <creds>` for Bundler < 2

Closes #9282 

## Context:

When a hostRule is defined such as the following: 
```
{
  "hostRules": [
    {
      "hostName": "private-registry.company.com",
      "hostType": "bundler",
      "token": "abc123"
    }
  ]
}
```
then Renovate will try to inject the following variable into the env: `BUNDLE_PRIVATE-REGISTRY__COMPANY__COM`.  
This is invalid as a bash enviroment variable and will therefore not work.

This is solved by using the above mentioned commands as precommand to add the authentication to Bundler.

## Documentation (please check one with an [x])

- [ ] I have updated the documentation, or
- [X] No documentation update is required

Instead the documentation would now actually work as you have an example with `-` in there.

## How I've tested my work (please tick one)

I have verified these changes via:

- [ ] Code inspection only, or
- [X] Newly added unit tests, or
- [ ] No new tests but ran on a real repository, or
- [ ] Both unit tests + ran on a real repository

### CODE REVIEW

**Review by rarkins — DISMISSED** [review]

File: lib/manager/bundler/artifacts.ts [hunk 1]
```diff
@@ -121,16 +125,55 @@ export async function updateArtifacts(
...
     );
 
+    // Detect hosts with a hyphen '-' in the url.
+    // Those cannot be added with environment variables but need to be added
+    // with the bundler config
+    const bundlerHostRulesAuthCommands: string[] = bundlerHostRules.reduce(
+      (authCommands: string[], hostRule) => {
+        if (hostRule.resolvedHost.includes('-')) {
+          authCommands.push(
+            `${hostRule.resolvedHost} ${getAuthenticationHeaderValue(hostRule)}`
```
Comment (viceice) [inline lib/manager/bundler/artifacts.ts:147]:
```suggestion
            `${hostRule.resolvedHost} ${getAuthenticationHeaderValue(hostRule)}`
```

store `getAuthenticationHeaderValue(hostRule)` in a local const and reuse, so we don't need to compute twice?

Comment (Chumper) [reply lib/manager/bundler/artifacts.ts]: Done

File: lib/manager/bundler/artifacts.spec.ts [hunk 2]
```diff
@@ -255,6 +255,135 @@ describe('bundler.updateArtifacts()', () => {
       ).toMatchSnapshot();
       expect(execSnapshots).toMatchSnapshot();
     });
+    it('injects bundler host configuration as command with bundler < 2', async () => {
```
Comment (viceice) [inline lib/manager/bundler/artifacts.spec.ts:258]:
```suggestion

    it('injects bundler host configuration as command with bundler < 2', async () => {
```

readabillity

File: lib/manager/bundler/artifacts.spec.ts [hunk 3]
```diff
@@ -255,6 +255,135 @@ describe('bundler.updateArtifacts()', () => {
...
+            binarySource: BinarySource.Docker,
+            constraints: {
+              bundler: '1.2',
+            },
+          },
+        })
+      ).toMatchSnapshot();
+      expect(execSnapshots).toMatchSnapshot();
+    });
+    it('injects bundler host configuration as command with bundler >= 2', async () => {
```
Comment (viceice) [inline lib/manager/bundler/artifacts.spec.ts:302]:
```suggestion

    it('injects bundler host configuration as command with bundler >= 2', async () => {
```

File: lib/manager/bundler/artifacts.spec.ts [hunk 4]
```diff
@@ -255,6 +255,135 @@ describe('bundler.updateArtifacts()', () => {
...
+            binarySource: BinarySource.Docker,
+            constraints: {
+              bundler: '2.1',
+            },
+          },
+        })
+      ).toMatchSnapshot();
+      expect(execSnapshots).toMatchSnapshot();
+    });
+    it('injects bundler host configuration as command with bundler == latest', async () => {
```
Comment (viceice) [inline lib/manager/bundler/artifacts.spec.ts:346]:
```suggestion

    it('injects bundler host configuration as command with bundler == latest', async () => {
```

File: lib/manager/bundler/artifacts.spec.ts [hunk 5]
```diff
@@ -255,6 +255,135 @@ describe('bundler.updateArtifacts()', () => {
...
+          config: {
+            ...config,
+            binarySource: BinarySource.Docker,
+          },
+        })
+      ).toMatchSnapshot();
+      expect(execSnapshots).toMatchSnapshot();
+    });
   });
   it('returns error when failing in lockFileMaintenance true', async () => {
```
Comment (viceice) [inline lib/manager/bundler/artifacts.spec.ts:388]:
```suggestion

  it('returns error when failing in lockFileMaintenance true', async () => {
```

File: lib/manager/bundler/artifacts.ts [hunk 6]
```diff
@@ -121,16 +125,54 @@ export async function updateArtifacts(
...
+          (authCommand) => `bundler config --local ${authCommand}`
+        )
+      );
+    } else if (bundlerHostRulesAuthCommands) {
+      preCommands.push(
+        ...bundlerHostRulesAuthCommands.map(
+          (authCommand) => `bundler config set --local ${authCommand}`
+        )
+      );
+    }
```
Comment (viceice) [inline lib/manager/bundler/artifacts.ts:174]: We should move those code to separate  file in a later pr.

**Review by viceice — APPROVED** [review]
LGTM, @rarkins What do you think?

**Pull request merged**